sdk/core/Azure.Core/src/Shared/DictionaryHeaders.cs

```diff
@@ -0,0 +1,134 @@
+        /// Adds a header value to the header collection.
+        /// </summary>
+        /// <param name="name">The header name.</param>
+        /// <param name="value">The header value.</param>
+        public void AddHeader(string name, string value)
+        {
+            if (!_headers.TryGetValue(name, out object objValue))
+            {
+                _headers[name] = value;
+            }
+            else
+            {
+                if (objValue is not List<string> values)
+                {
+                    _headers[name] = values = new List<string> { objValue as string, value };
+                }
+                else
+                {
+                    values.Add(value);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns header value if the header is stored in the collection. If the header has multiple values they are going to be joined with a comma.
+        /// </summary>
+        /// <param name="name">The header name.</param>
+        /// <param name="value">The reference to populate with value.</param>
+        /// <returns><c>true</c> if the specified header is stored in the collection, otherwise <c>false</c>.</returns>
+        public bool TryGetHeader(string name, out string value)
+        {
+            if (_headers.TryGetValue(name, out object objValue))
+            {
+                if (objValue is List<string> values)
+                {
+                    value = JoinHeaderValue(values);
+                }
+                else
+                {
+                    value = objValue as string;
+                }
+                return true;
+            }
+
+            value = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Returns header values if the header is stored in the collection.
+        /// </summary>
+        /// <param name="name">The header name.</param>
+        /// <param name="values">The reference to populate with values.</param>
+        /// <returns><c>true</c> if the specified header is stored in the collection, otherwise <c>false</c>.</returns>
+        public bool TryGetHeaderValues(string name, out IEnumerable<string> values)
+        {
+            var result = _headers.TryGetValue(name, out object objValue);
+            if (objValue is List<string> valuesList)
+            {
+                values = valuesList;
+            }
+            else
+            {
+                values = new List<string> { objValue as string };
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Returns <c>true</c> if the header is stored in the collection.
+        /// </summary>
+        /// <param name="name">The header name.</param>
+        /// <returns><c>true</c> if the specified header is stored in the collection, otherwise <c>false</c>.</returns>
+        public bool ContainsHeader(string name)
+        {
+            return TryGetHeaderValues(name, out _);
```

Comment:
If we are optimizing let's avoid a string allocation here and just do ContainsKey